File: android/library/core/src/main/java/com/liferay/mobile/screens/assetlist/AssetListScreenlet.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.mobile.screens.assetlist;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;

import com.liferay.mobile.screens.R;
import com.liferay.mobile.screens.assetlist.interactor.AssetListInteractor;
import com.liferay.mobile.screens.assetlist.interactor.AssetListInteractorImpl;
import com.liferay.mobile.screens.assetlist.interactor.AssetListInteractorListener;
import com.liferay.mobile.screens.base.list.BaseListScreenlet;
import com.liferay.mobile.screens.cache.OfflinePolicy;
import com.liferay.mobile.screens.context.LiferayServerContext;

import java.util.Locale;

/**
 * @author Silvio Santos
 */
public class AssetListScreenlet
	extends BaseListScreenlet<AssetEntry, AssetListInteractor>
	implements AssetListInteractorListener {

	public AssetListScreenlet(Context context) {
		super(context);
	}

	public AssetListScreenlet(Context context, AttributeSet attributes) {
		super(context, attributes);
	}

	public AssetListScreenlet(Context context, AttributeSet attributes, int defaultStyle) {
		super(context, attributes, defaultStyle);
	}

	public long getClassNameId() {
		return _classNameId;
	}

	public void setClassNameId(long classNameId) {
		_classNameId = classNameId;
	}

	public long getGroupId() {
		return _groupId;
	}

	public void setGroupId(long groupId) {
		_groupId = groupId;
	}

	public OfflinePolicy getOfflinePolicy() {
		return _offlinePolicy;
	}

	public void setOfflinePolicy(OfflinePolicy offlinePolicy) {
		_offlinePolicy = offlinePolicy;
	}

	public String getPortletItemName() {
		return _portletItemName;
	}

	public void setPortletItemName(String portletItemName) {
		_portletItemName = portletItemName;
	}

	@Override
	public void loadingFromCache(boolean success) {
		if (getListener() != null) {
			getListener().loadingFromCache(success);
		}
	}

	@Override
	public void retrievingOnline(boolean triedInCache, Exception e) {
		if (getListener() != null) {
			getListener().retrievingOnline(triedInCache, e);
		}
	}

	@Override
	public void storingToCache(Object object) {
		if (getListener() != null) {
			getListener().storingToCache(object);
		}
	}

	@Override
	protected void loadRows(AssetListInteractor interactor, int startRow, int endRow, Locale locale)
		throws Exception {

		interactor.loadRows(_groupId, _classNameId, _portletItemName, startRow, endRow, locale);
	}

	@Override
	protected View createScreenletView(Context context, AttributeSet attributes) {
		TypedArray typedArray = context.getTheme().obtainStyledAttributes(
			attributes, R.styleable.AssetListScreenlet, 0, 0);
		_classNameId = typedArray.getInt(
			R.styleable.AssetListScreenlet_classNameId, 0);

		int offlinePolicy = typedArray.getInt(R.styleable.WebContentDisplayScreenlet_offlinePolicy,
			OfflinePolicy.REMOTE_ONLY.ordinal());
		_offlinePolicy = OfflinePolicy.values()[offlinePolicy];

		long groupId = LiferayServerContext.getGroupId();

		_groupId = castToLongOrUseDefault(typedArray.getString(
			R.styleable.AssetListScreenlet_groupId), groupId);

		_portletItemName = typedArray.getString(R.styleable.AssetListScreenlet_portletItemName);

		typedArray.recycle();

		return super.createScreenletView(context, attributes);
	}

	@Override
	protected AssetListInteractor createInteractor(String actionName) {
		return new AssetListInteractorImpl(getScreenletId(), _offlinePolicy);
	}

	private OfflinePolicy _offlinePolicy;
	private long _classNameId;
	private long _groupId;
	private String _portletItemName;

}

File: ios/Library/Core/AssetListScreenlet/AssetListScreenlet.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit


@objc public protocol AssetListScreenletDelegate {

	optional func onAssetListResponse(entries:[AssetListScreenletEntry])
	optional func onAssetListError(error: NSError)

	optional func onAssetSelected(entry: AssetListScreenletEntry)

}


@objc public class AssetListScreenletEntry : NSObject {

	public let attributes:[String:AnyObject]

	public var title: String {
		return attributes["title"] as String
	}

	override public var description: String {
		return attributes["description"] as String
	}

	public var classNameId: Int64 {
		return Int64(attributes["classNameId"] as Int)
	}

	public var classPK: Int64 {
		return Int64(attributes["classPK"] as Int)
	}

	public var groupId: Int64 {
		return Int64(attributes["groupId"] as Int)
	}

	public var companyId: Int64 {
		return Int64(attributes["companyId"] as Int)
	}

	public var entryId: Int64 {
		return Int64(attributes["entryId"] as Int)
	}


	//MARK: Init

	public init(attributes:[String:AnyObject]) {
		self.attributes = attributes
	}

}


@IBDesignable public class AssetListScreenlet: BaseListScreenlet {

	@IBInspectable public var groupId: Int64 = 0
	@IBInspectable public var classNameId: Int = 0

	@IBOutlet public var delegate: AssetListScreenletDelegate?

	//MARK: BaseListScreenlet

	override internal func createPaginationOperation(
			#page: Int,
			computeRowCount: Bool)
			-> LiferayPaginationOperation {

		let operation = LiferayAssetListPageOperation(
				screenlet: self,
				page: page,
				computeRowCount: computeRowCount)

		operation.groupId = (self.groupId != 0)
				? self.groupId : LiferayServerContext.groupId

		operation.classNameId = Int64(self.classNameId)

		return operation
	}

	override internal func convert(#serverResult:[String:AnyObject]) -> AnyObject {
		return AssetListScreenletEntry(attributes: serverResult)
	}

	override internal func onLoadPageError(#page: Int, error: NSError) {
		super.onLoadPageError(page: page, error: error)

		delegate?.onAssetListError?(error)
	}

	override internal func onLoadPageResult(
			#page: Int,
			serverRows: [[String:AnyObject]],
			rowCount: Int)
			-> [AnyObject] {

		let rowObjects = super.onLoadPageResult(
				page: page,
				serverRows: serverRows,
				rowCount: rowCount)

		let assetEntries = rowObjects.map() { $0 as AssetListScreenletEntry }

		delegate?.onAssetListResponse?(assetEntries)

		return rowObjects
	}

	override internal func onSelectedRow(row: AnyObject) {
		delegate?.onAssetSelected?(row as AssetListScreenletEntry)
	}

}
